### PR TITLE
Kfs 1330/don't store secrets in history

### PR DESCRIPTION
**Pull request by Guttz**
Release Notes
-------------

Checklist
---------
- [x] Leave this box unchecked if features are not yet available in production

What
----

This will avoid storing secrets in the history. This will be important with the coming up integrations we're adding.

References
----------

Test & Review
-------------

yes, see tests

### CODE REVIEW

File: pkg/flink/internal/store/store_utils.go [hunk 1]
```diff
@@ -388,6 +391,25 @@ func startsWithValidSQL(statement string) bool {
...
+	count := len(parts)
+
+	if count < 1 {
+		return false
+	}
+
+	last := strings.ToLower(parts[count-1])
+
+	// add *ecrets, s*crets and so on. Maybe not * but only two letter that are wrong after the s? hmm
+	matches, err := regexp.MatchString(".*?ecret|s.{0,2}?cret|se.{0,2}?ret|sec.{0,2}?et|secr.{0,2}?t|secre.{0,2}?", last)
```
Comment (swist) [inline pkg/flink/internal/store/store_utils.go:405]: can you see if there's a levenshtein distance algo somewhere for go (or just build it yourself as a util). Then you'll also cover mishits on keyboard. Let's say score of 2 is what you want

**Review by swist — CHANGES_REQUESTED** [review]

File: pkg/flink/internal/store/store_utils.go (hunk 1 above)
Comment (Guttz) [reply pkg/flink/internal/store/store_utils.go:405]: lemme look into that

Comment (Guttz) [reply pkg/flink/internal/store/store_utils.go]: Neat pointer. Thanks!

**Comment by brianstrauch** [comment]
1. If I type a secret in the terminal it gets recorded in ~/.bash_history or ~/.zsh_history. Much better solutions involve reading secrets from stdin or a file. Have we considered those?
2. Edit distance seems pretty arbitrary here. What if I mistype by three characters?

File: pkg/flink/config/local_statements.go [hunk 2]
```diff
@@ -19,4 +19,7 @@ const (
 	KeyResultsTimeout = "client.results-timeout"
 	KeyServiceAccount = "client.service-account"
 	KeyStatementName  = "client.statement-name"
+	KeyFlinkSecret    = "confluent.user.flink.secret"
 )
+
+var Sensitivekeys = []string{KeyFlinkSecret}
```
Comment (sgagniere) [inline pkg/flink/config/local_statements.go:25]:
```suggestion
var SensitiveKeys = []string{KeyFlinkSecret}
```

**Review by sgagniere — APPROVED** [review]
Some small nitpicks.

File: pkg/flink/internal/store/store_utils.go [hunk 3]
```diff
@@ -70,14 +73,20 @@ func (s *Store) processSetStatement(statement string) (*types.ProcessedStatement
 			Suggestion: `please provide a non-empty statement name with "SET 'client.statement-name'='non-empty-name'"`,
 		}
 	}
+
+	hasSensitiveKey := lo.SomeBy(config.Sensitivekeys, func(sensitiveKey string) bool {
```
Comment (sgagniere) [inline pkg/flink/internal/store/store_utils.go:77]:
```suggestion
	hasSensitiveKey := lo.SomeBy(config.SensitiveKeys, func(sensitiveKey string) bool {
```

**Review by sgagniere — APPROVED** [review]

File: pkg/flink/internal/store/store_utils.go [hunk 4]
```diff
@@ -388,6 +397,13 @@ func startsWithValidSQL(statement string) bool {
 	return config.SQLKeywords.Contains(firstWord)
 }
 
+func isKeySimilarToSensitiveKey(sensitiveKeyName string, key string) bool {
+	key = strings.ToLower(key)
+	distance := levenshtein.DistanceForStrings([]rune(sensitiveKeyName), []rune(key), levenshtein.DefaultOptions)
+
+	return distance <= 2
```
Comment (pgaref) [inline pkg/flink/internal/store/store_utils.go:404]: I would even go higher ere but no strong feelings :)

File: pkg/flink/config/local_statements.go [hunk 5]
```diff
@@ -19,4 +19,7 @@ const (
 	KeyResultsTimeout = "client.results-timeout"
 	KeyServiceAccount = "client.service-account"
 	KeyStatementName  = "client.statement-name"
+	KeyFlinkSecret    = "confluent.user.flink.secret"
 )
+
+var SensitiveKeys = []string{KeyFlinkSecret}
```
Comment (pgaref) [inline pkg/flink/config/local_statements.go:25]: Any more keys we should probably add here? Maybe @twalthr ?

**Pull request merged**